### PR TITLE
feat(activity): add activity 2

### PR DESCRIPTION
**Pull request by lwasser**
This is the second activity in the workshop. It's complete short of the answers to the OYO sections. i have an answer key script as well that i can push separately.

### CODE REVIEW

**Comment by lwasser** [comment]
NOTE: the CI issues will be fixed when some of the other pr's are merged. I'm worrying less about that now and more about content!!

File: clean-modular-code/activity-2/clean-code-activity-2.md [hunk 1]
```diff
@@ -11,22 +11,37 @@ kernelspec:
...
-    * Then clearly document what the document does to help your future self or another user understand its intent.
-* REMOVE - part 3 Create loops that iterate over repetitive tasks.
-* Use conditional statements to control if and when code is executed.
+You can use the script you worked on in Activity 1 here. Or you can use the cleaned-up script provided below. 
+
++++ {"editable": true, "slideshow": {"slide_type": ""}}
+
+## Remember the strategies for cleaner code 
+
+* Document what the code does by adding a docstring at the top of the script to help your future self, a future lab mate, or another user understand that tasks that your code performs.
```
Comment (ucodery) [inline clean-modular-code/activity-2/clean-code-activity-2.md:31]:
```suggestion
* Document what the code does by adding a docstring at the top of the script to help your future self, a future lab mate, or another user understand the tasks that your code performs.
```

File: clean-modular-code/activity-2/clean-code-activity-2.md [hunk 2]
```diff
@@ -36,15 +51,19 @@ What tools could make the code more DRY?
 
 ## Reproducibility: will this code run on other machines?
 
-One big challenge when considering the reproducibility of workflows is ensuring that your code runs seamlessly on different machines or environments. One key issue is file paths. Hard-coded paths, like "data/part-1-data.json", can lead to errors if the directory structure on another machine differs from your setup. Paths should be constructed dynamically, using tools like Python’s pathlib or environment variables to ensure flexibility across different systems.
+One big challenge when considering the reproducibility of workflows is ensuring that your code runs seamlessly on different machines or environments. One key issue is file paths. Hard-coded paths, like `data/part-1-data.json`, can lead to errors if the directory structure on another machine differs from your setup. Paths should be constructed dynamically, using tools like Python’s `pathlib` or environment variables to ensure flexibility across different systems.
```
Comment (ucodery) [inline clean-modular-code/activity-2/clean-code-activity-2.md:54]: I definitely thought you were leading up to Window vs POSIX path slashes. I think that is a more basic, and easier to automate, difference between environments. If the directory structure actually changes in new environments, I think there is only so much you can do without needing user input.

File: clean-modular-code/activity-2/clean-code-activity-2.md [hunk 3]
```diff
@@ -157,22 +190,36 @@ combined_papers_df.shape
...
 
-If you get through the above activity with time to spare, here are a few additional tasks that you can work on. 
+If you complete the above activity with time to spare, here are a few additional tasks you can work on. 
 
-1. In this file, the package titles can be found in the title column, as follows:
+In the JSON data that you have been processing, the package titles can be found in the title column as follows: `"package-name: title here."` 
 
-"package-name: title here." 
+1. Add a processing step to the code that pulls the package name out from the title column and creates a new column called `package-name`.
+1. Create a workflow that calculates how many packages in that list are pyOpenSci packages. To calculate how many are pyOpenSci packages,   you will need the `pyos-joss-accepted-packages.csv` file in the `data/` directory with the activity 2 JSON data. This .csv file lists all pyOpenSci packages that have also gone through JOSS review.
```
Comment (ucodery) [inline clean-modular-code/activity-2/clean-code-activity-2.md:202]:
```suggestion
1. Create a workflow that calculates how many packages in that list are pyOpenSci packages. To calculate how many are pyOpenSci packages, you will need the `pyos-joss-accepted-packages.csv` file in the `data/` directory with the activity 2 JSON data. This .csv file lists all pyOpenSci packages that have also gone through JOSS review.
```

File: clean-modular-code/activity-2/clean-code-activity-2.md (hunk 2 above)
Comment (lwasser) [reply clean-modular-code/activity-2/clean-code-activity-2.md:54]: Oh - @ucodery i didn't understand this comment at first but now i think i do. 

this section IS about windows vs POSIC path slashes. I think my language that says "directory structure on another machine differs" is confusing / incorrect!!

that is a much harder problem that we can't solve with code as you point out. How about i edit this paragraph to make it more clear. it's fully about path creation which we can handle with code.

**Comment by lwasser** [comment]
Cool - CI is failing with things are are fixed in #18 #13  i am going to merge this soon after one last round of revisions and adding answers to the OYOs

**Review by ucodery — APPROVED** [review]
I just read this one section, I couldn't easily tell if anything else has changed

File: clean-modular-code/activity-2/clean-code-activity-2.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+## To begin - identify redundant parts of this code 
+
+Create pseudocode (or text) describing the process steps represented in the code below. How could the code be organized differently and more efficiently? 
+
+What tools could make the code more DRY?
+
+## Reproducibility: will your code run on other machines?
+
+When considering workflow reproducibility, a key challenge is ensuring that your code runs seamlessly across different machines. Hard-coded paths, like data/part-1-data.json, can cause errors if created on a Windows machine but run on a Mac (Unix-based/POSIX) machine.
```
[the remark this answers is not in the document]
Comment (lwasser) [reply clean-modular-code/activity-2/clean-code-activity-2.md:54]: oh you're right! that's an important nit!!

**Comment by lwasser** [comment]
thank you, @ucodery. If you want to have a look at a preview of the published (ci) page - it's here - https://output.circle-artifacts.com/output/job/b18a4bf1-8f83-4764-a52b-e0e61a07f909/artifacts/0/html/clean-modular-code/activity-2/clean-code-activity-2.html I've made small edits and clarified things which is why it may be hard to identify what has changed.

The big change here is that I rewrote the POSIX path section and added examples. I'm hopeful that makes it more clear!  Thank you for the review comments - things were not worded well before.

**Comment by lwasser** [comment]
I am going to merge this PR, rebase and focus more time on activity 3 next week. Thank you again @ucodery for all of the input!! the CI issues are related to other lessons not in this PR. They will all get fixed once things are merged! Lesson dev is iterative as we focus on the content and move things around accordingly.

**Pull request merged**